sktalk/corpus/utterance.py

```diff
@@ -12,34 +14,31 @@ class Utterance:
     participant: Optional[str] = None
     time: Optional[list] = None
     begin: Optional[str] = None
+    begin_timestamp: Optional[str] = None
     end: Optional[str] = None
-    metadata: Optional[dict[str, Any]] = None
-    utterance_clean: Optional[str] = None
+    end_timestamp: Optional[str] = None
+    utterance_raw: Optional[str] = None
     utterance_list: Optional[list[str]] = None
     n_words: Optional[int] = None
     n_characters: Optional[int] = None
-    time_to_next: Optional[int] = None
-    dyadic: Optional[bool] = None
     FTO: Optional[int] = None
+    metadata: Optional[dict[str, Any]] = None
 
     def __post_init__(self):
-        # clean utterance:
-        if not self.utterance_clean:
-            self._clean_utterance()
-
-        # generate a list of words in the utterance
-        if not self.utterance_list:
-            self.utterance_list = self.utterance_clean.split()
+        if not self.utterance_raw:  # if reading in existing data, we do not want to overwrite the raw utterance
+            self.utterance_raw = self.utterance
+        self.utterance = self._clean_utterance(self.utterance)
+        self.utterance_list = self.utterance.split()
+        self.n_words = len(self.utterance_list)
+        self.n_characters = sum(len(word) for word in self.utterance_list)
 
-        # count words and characters
-        if not self.n_words:
-            self.n_words = len(self.utterance_list)
-        if not self.n_characters:
-            self.n_characters = sum(len(word) for word in self.utterance_list)
+        self._validate_time()
 
-        # calculate timestamps
-        if not self.begin or not self.end:
-            self._split_time()
+        if (not self.begin or not self.end) and self.time:
+            self.begin = self.time[0]
+            self.end = self.time[1]
+            self.begin_timestamp = self._to_timestamp(self.begin)
+            self.end_timestamp = self._to_timestamp(self.end)
```

Comment:
Perhaps aggregate all the time-related post-processing and validation in one method (e.g. `_process_time()`)?

Reply:
I thought about this — it was actually the case before. But I like some clarity about which fields are updated after the class is instantiated, and putting this in a single function would remove that.